dali/operators/reader/loader/video/video_file.cc

```diff
@@ -0,0 +1,232 @@
+
+  DALI_ENFORCE(
+    avcodec_open2(codec_ctx_, codec_, nullptr) == 0,
+    "Could not initialize codec context");
+
+  frame_ = av_frame_alloc();
+  DALI_ENFORCE(frame_, "Could not allocate the av frame");
+
+  packet_ = av_packet_alloc();
+  DALI_ENFORCE(packet_, "Could not allocate av packet");
+}
+
+void VideoFileCPU::FindVideoStream() {
+  for (size_t i = 0; i < ctx_->nb_streams; ++i) {
+    auto stream = ctx_->streams[i];
+    codec_params_ = ctx_->streams[i]->codecpar;
+    codec_ = avcodec_find_decoder(codec_params_->codec_id);
+
+    if (codec_ == nullptr) {
+      continue;
+    }
+
+    if (codec_->type == AVMEDIA_TYPE_VIDEO) {
+        stream_id_ = i;
+        return;
+    }
+  }
+
+  DALI_FAIL(make_string("Could not find a valid video stream in file ", filename_));
+}
+
+VideoFileCPU::VideoFileCPU(std::string &filename) : 
+  filename_(filename) {
+  ctx_ = avformat_alloc_context();
+  DALI_ENFORCE(ctx_, "Could not create avformat context");
+
+  DALI_ENFORCE(
+    avformat_open_input(&ctx_, filename.c_str(), nullptr, nullptr) == 0,
+    make_string("Failed to open video file at path ", filename));
+
+  FindVideoStream();
+  InitAvState();
+  BuildIndex();
+}
+
+void VideoFileCPU::BuildIndex() {
+  // TODO(awolant): Optimize this function for: 
+  //  - CFR
+  //  - index present in the header
+
+  int last_keyframe = -1;
+  while (ReadRegularFrame(nullptr, false)) {
+    IndexEntry entry;
+    entry.is_keyframe = frame_->key_frame;
+    entry.pts = frame_->pts;
+    entry.is_flush_frame = false;
+    
+    if (entry.is_keyframe) {
+      last_keyframe = index_.size();
+    }
+    entry.last_keyframe_id = last_keyframe;
+    index_.push_back(entry);
+  }
+  while(ReadFlushFrame(nullptr, false)) {
+    IndexEntry entry;
+    entry.is_keyframe = false;
+    entry.pts = frame_->pts;
+    entry.is_flush_frame = true;
+    entry.last_keyframe_id = last_keyframe;
+    index_.push_back(entry);
+  }
+  Reset();
+}
+
+void VideoFileCPU::CopyToOutput(uint8_t *data) {
+  dest_[0] = data;
+  dest_linesize_[0] = frame_->width * Channels();
+  if (sws_scale(sws_ctx_, frame_->data, frame_->linesize, 0, frame_->height, dest_, dest_linesize_) < 0) {
+    DALI_FAIL("Could not convert frame data to RGB");
+  }
+}
+
+bool VideoFileCPU::ReadRegularFrame(uint8_t *data, bool copy_to_output) {
+  while(av_read_frame(ctx_, packet_) >= 0) {
+    if (packet_->stream_index != stream_id_) {
+      continue;
+    }
+
+    if (avcodec_send_packet(codec_ctx_, packet_) < 0) {
+      DALI_FAIL("Failed to send packet to decoder");
+    }
+
+    int ret = avcodec_receive_frame(codec_ctx_, frame_);
+
+    if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
+      continue;
+    }
+
+    if (!copy_to_output) {
+      return true;
+    }
+
+    if (sws_ctx_ == nullptr) {
+      sws_ctx_ = sws_getContext(
+        Width(), 
+        Height(), 
+        codec_ctx_->pix_fmt, 
+        Width(), 
+        Height(), 
+        AV_PIX_FMT_RGB24, 
+        SWS_BILINEAR, 
```

Comment:
What other interpolations we may select? Maybe it should be adjustable?

Reply:
Yes, we can use multiple other interpolations and output formats. This is also an entry point to `libsw` scaling mechanisms. This will allow to do better version of `VideoReaderResize`.
Both will be added in the future PRs as right now I want to focus on making it working end-to-end in one scenario. Then I'll add more options like these.